cmd/ncproxy/ncproxy.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/configagent"
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/ncproxygrpc"
+	"github.com/Microsoft/hcsshim/internal/computeagent"
+	"github.com/Microsoft/hcsshim/internal/hcsoci"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/ncproxyttrpc"
+	"github.com/containerd/ttrpc"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+// GRPC service exposed for use by a network configuration agent. Holds a mutex for
+// updating map of client connections.
+type grpcService struct {
+	m sync.Mutex
+}
+
+func (s *grpcService) AddNIC(ctx context.Context, req *ncproxygrpc.AddNICRequest) (*ncproxygrpc.AddNICResponse, error) {
+	log.G(ctx).WithFields(logrus.Fields{
+		"namespace":  req.NamespaceID,
+		"endpointID": req.EndpointID,
+		"nicID":      req.NicID,
+	}).Info("AddNIC request")
+
+	if client, ok := namespaceToShim[req.NamespaceID]; ok {
+		caReq := &computeagent.AddNICRequest{
+			NamespaceID: req.NamespaceID,
+			NicID:       req.NicID,
+			EndpointID:  req.EndpointID,
+		}
+		if _, err := client.AddNIC(ctx, caReq); err != nil {
+			return nil, err
+		}
+		return &ncproxygrpc.AddNICResponse{}, nil
+	}
+	return nil, fmt.Errorf("no shim registered for namespace %s", req.NamespaceID)
+}
+
+func (s *grpcService) RegisterNetworkConfigAgent(ctx context.Context, req *ncproxygrpc.RegisterNetworkConfigAgentRequest) (*ncproxygrpc.RegisterNetworkConfigAgentResponse, error) {
+	log.G(ctx).WithFields(logrus.Fields{
+		"networkID":    req.NetworkID,
+		"agentAddress": req.AgentAddress,
+	}).Info("RegisterNetworkConfigAgent request")
+
+	// TODO (Dcantah): Make timeout configurable
```

Comment:
maybe continue until success or context deadline if available, otherwise a default of 5 seconds?

Reply:
I was just thinking a config file field, and if not set then 5 second default.